File: Homework-by-C++/Homework-Week-1/Hw 1_4 全排列.cpp
#include <iostream>
#include <vector>
#include <set>
#include <algorithm>

using namespace std;

void generate_permutations(vector<int>& arr, int n, vector<int>& curr_permutation) {
    if (curr_permutation.size() == n) {
        for (int i = 0; i < n; i++) {
            cout << curr_permutation[i];
            if (i < n - 1) cout << " ";
        }
        cout << endl;
        return;
    }

    set<int> seen;
    for (int i = 0; i < arr.size(); i++) {
        if (seen.find(arr[i]) == seen.end()) {
            seen.insert(arr[i]);
            curr_permutation.push_back(arr[i]);
            vector<int> next_arr = arr;
            next_arr.erase(next_arr.begin() + i);
            generate_permutations(next_arr, n, curr_permutation);
            curr_permutation.pop_back();
        }
    }
}

int main() {
    int N;
    cin >> N;
    vector<int> arr(N);
    for (int i = 0; i < N; ++i) {
        cin >> arr[i];
    }

    vector<int> curr_permutation;
    generate_permutations(arr, N, curr_permutation);

    system("pause");
    return 0;


}

File: Homework-by-python/Homework-Week-1/Hw 1_4 全排列.py
def generate_permutations(arr, n, curr_permutation):
    if len(curr_permutation) == n:
        print(" ".join(map(str, curr_permutation)))
        return
    
    seen = set()
    for i in range(len(arr)):
        if arr[i] not in seen:
            seen.add(arr[i])
            curr_permutation.append(arr[i])
            next_arr = arr[:i] + arr[i+1:]
            generate_permutations(next_arr, n, curr_permutation)
            curr_permutation.pop()

N = int(input())
arr = list(map(int, input().split()))

generate_permutations(arr, N, [])
